pkg/monitor/intervalcreation/podlogs.go

```diff
@@ -0,0 +1,145 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+// SubStringLevel defines a sub-string we'll scan pod log lines for, and the level the resulting
+// interval should have. (Info, Warning, Error)
+type SubStringLevel struct {
+	subString string
+	level     monitorapi.EventLevel
+}
+
+type PodLogIntervalGenerator struct {
+	namespace string
+	// selector is a label selector for which pods to gather from. (i.e. app=etcd)
+	selector  string
+	container string
+	// subStrings are matched against every log line to see what we should generate an interval for.
+	subStrings []SubStringLevel
+	// lineParser is called to convert a log line to an EventInterval. Function is only called if
+	// the line matches one of the substrings.
+	lineParser func(locator, line string, intervalLevel monitorapi.EventLevel, logger logrus.FieldLogger) (*monitorapi.EventInterval, error)
+}
+
+func (g PodLogIntervalGenerator) ScanLine(pod *kapiv1.Pod, line string, beginning, end time.Time, logger logrus.FieldLogger) (*monitorapi.EventInterval, error) {
+	for _, subStr := range g.subStrings {
+		if strings.Contains(line, subStr.subString) {
+			locator := monitorapi.LocatePodContainer(pod, g.container)
+			// Add a src/podLog to the locator for filtering:
+			locator = fmt.Sprintf("%s src/podLog", locator)
+			interval, err := g.lineParser(locator, line, subStr.level, logger)
+			if err != nil {
+				return nil, err
+			}
+			// If we're outside our beginning/end times, we're throwing this interval away.
+			if interval.From.Before(beginning) || interval.From.After(end) ||
+				interval.To.Before(beginning) || interval.To.After(end) {
+				return nil, nil
+			}
+			return interval, err
+		}
+	}
+	return nil, nil
+}
+
+type etcdLogLine struct {
+	Level     string    `json:"level"`
+	Timestamp time.Time `json:"ts"`
+	Msg       string    `json:"msg"`
+}
+
+// etcdLogParser handles etcd logs which are already nicely json formatted such as:
+//
+// {"level":"info","ts":"2023-03-03T18:09:01.471Z","caller":"mvcc/index.go:214","msg":"compact tree index","revision":738215}
+func etcdLogParser(locator, line string, level monitorapi.EventLevel, logger logrus.FieldLogger) (*monitorapi.EventInterval, error) {
+	parsedLine := etcdLogLine{}
+	err := json.Unmarshal([]byte(line), &parsedLine)
+	if err != nil {
+		logger.WithError(err).Errorf("error parsing matched log line: %s", line)
+		return nil, err
+	}
+	return &monitorapi.EventInterval{
+		Condition: monitorapi.Condition{
+			Level:   level,
+			Locator: locator,
+			Message: parsedLine.Msg,
+		},
+		From: parsedLine.Timestamp,
+		To:   parsedLine.Timestamp.Add(1 * time.Second),
+	}, nil
+}
+
+func buildLogGatherers() []PodLogIntervalGenerator {
+	return []PodLogIntervalGenerator{
+		{
+			namespace: "openshift-etcd",
+			selector:  "app=etcd",
+			container: "etcd",
+			subStrings: []SubStringLevel{
+				{"slow fdatasync", monitorapi.Warning},
+			},
+			lineParser: etcdLogParser,
+		},
+	}
+}
+
```

Comment:
Presume this is doing the interval extraction, are we losing the fact that it is 'slow fdatasync' related in the interval chart or am I missing how the interval is generated?

![image](https://user-images.githubusercontent.com/36795216/224127780-4f1c109e-7c0f-46b1-8eca-775a921aaa68.png)

Reply:
That is where it would show up if it happened, hopefully you are looking at older job runs prior to most recent commits, I had a couple common strings in there to ensure I was getting intervals as the slow sync one is relatively rare in comparison.

Reply:
Heh, that is likely it.  I followed an outdated [link](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/27772/pull-ci-openshift-origin-master-e2e-aws-ovn-etcd-scaling/1633489607620300800) from that darn trtbot ;-)

Reply:
Had to zoom in but https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/27772/pull-ci-openshift-origin-master-e2e-azure-ovn-etcd-scaling/1633896651523362816 had it.